packages/eslint-plugin/rules/no-unused-vars-before-return.js

```diff
@@ -0,0 +1,55 @@
+module.exports = {
+	meta: {
+		type: 'problem',
+		schema: [],
+	},
+	create( context ) {
+		return {
+			ReturnStatement( node ) {
+				let functionScope = context.getScope();
+				while ( functionScope.type !== 'function' && functionScope.upper ) {
+					functionScope = functionScope.upper;
+				}
+
+				if ( ! functionScope ) {
+					return;
+				}
+
+				for ( const variable of functionScope.variables ) {
+					const isAssignmentCandidate = variable.defs.some( ( def ) => {
+						return (
+							def.node.type === 'VariableDeclarator' &&
+							// Allow declarations which are not initialized.
+							def.node.init &&
+							// Target function calls as "expensive".
+							def.node.init.type === 'CallExpression' &&
+							// Allow unused if part of an object destructuring.
+							def.node.id.type !== 'ObjectPattern' &&
+							// Only target assignments preceding `return`.
+							def.node.end < node.end
+						);
+					} );
+
+					if ( ! isAssignmentCandidate ) {
+						continue;
+					}
+
+					// The first entry in `references` is the declaration
+					// itself, which can be ignored.
+					const isUsedBeforeReturn = variable.references.slice( 1 ).some( ( reference ) => {
+						return reference.identifier.end < node.end;
+					} );
+
+					if ( isUsedBeforeReturn ) {
+						continue;
+					}
+
+					context.report(
+						node,
+						`Declared variable \`${ variable.name }\` is unused before a return path`
```

Comment:
I'm not quite sure that is going to be clear enough when printed on the console. Should it include more detailed info that explains that there is a return path where it is not used before it is even used? I doesn't read good but I hope it explain my concern. Anyway, I don't find it as blocker :)

Reply:
I played with the message a bit more. One thing which confused me is its placement on the `return` statement vs. the assignment itself, which I've found can be adjusted. Though there's also some value in having it on the actual return which flags it as offending 🤷‍♂️ 

Any thoughts on (attached to the variable declaration):

>Variables should not be assigned until just prior its first reference. An early return statement may leave this variable unused.

Reply:
This is much better. Let's proceed with it 👍